### ISSUE TITLE
CUDA error: an illegal memory access was encountered

### ISSUE THREAD

**Issue by RuiWang1998** [HazyResearch/flash-attention#124]
Hi,

Thanks for this great work!

We have managed to get it working on one node (8 x A100 80GB) with PyTorch's FSDP and it runs very fast and very efficiently!

However, when we try to scale it up to two nodes, this occurs:

```  File "/home/rui/code/OmegaPLM/training/trainer.py", line 979, in train_step
    loss.backward()
  File "/home/rui/anaconda3/envs/t/lib/python3.9/site-packages/torch/_tensor.py", line 488, in backward
    torch.autograd.backward(
  File "/home/rui/anaconda3/envs/t/lib/python3.9/site-packages/torch/autograd/__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.```

We are packing our sequences directly to conform to the inputs required by `_flash_attn_forward` to save more memory and compute, if this makes any difference. and training a T5 model `cu_seqlens`.

This is one of the two issues that have shown up. Previously it was a `CUBLAS_STATUS_INTERNAL_ERROR` but somehow we were able to get rid of (not sure if this is true though).

Could you please give some advice?

**Comment by RuiWang1998** [comment]
And we are using `1.13.1` for torch and `11.7` for cuda just for context.

**Comment by tridao** [comment]
Thanks for the report.
To isolate the issue, can you run without packing the sequences (all sequences in the batch have the same length) and verify that there's no error there?

How are you packing the sequences? If the indices are wrong, it could point to non-existent memory and could cause memory error.
Here's an example of BERT [implementation](https://github.com/HazyResearch/flash-attention/blob/57ee618170e1adecbf787365cdf330c63768abd2/flash_attn/models/bert.py#L138) that removes all the padding tokens.

**Comment by RuiWang1998** [comment]
Hi @tridao ,

Thanks so much for your reply!

I think I have figured things out a bit. And it's two fold, both of which are direct results of the fact that even we are concatenating sequences, to avoid fragmentation, we force a pre-determined length i.e., padding:

1. The issue where we encountered cuda errors were because I passed in `q[:cut_off]` instead of `q`, where `cut_off` is where the padding started. I thought this should be fine as it did in other first-party ops from torch, torch, as it turned out, may have freed the other parts of the input `q`.
2. The `torch.empty` used in initializing the output of the function as well as the input gradient caused problems for our padding as those values are not necessarily defined, so we change those to `torch.zeros` and it has been working like a charm since!

Again, thanks for the great work!
